pkg/operator/latencyprofilecontroller/status_utils.go

```diff
@@ -0,0 +1,74 @@
+package latencyprofilecontroller
+
+import (
+	"context"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/util/retry"
+)
+
+const (
+	reasonWorkerLatencyProfileUpdated = "ProfileUpdated"
+	reasonWorkerLatencyProfileEmpty   = "ProfileEmpty"
+)
+
+// setWLPStatusCondition is used to set condition in config node object status.workerLatencyProfileStatus
+func setWLPStatusCondition(conditions *[]metav1.Condition, newCondition metav1.Condition) {
+	if conditions == nil {
+		conditions = &[]metav1.Condition{}
+	}
+	existingCondition := findWLPStatusCondition(*conditions, newCondition.Type)
+	if existingCondition == nil {
+		newCondition.LastTransitionTime = metav1.NewTime(time.Now())
+		*conditions = append(*conditions, newCondition)
+		return
+	}
+
+	if existingCondition.Status != newCondition.Status {
+		existingCondition.Status = newCondition.Status
+		existingCondition.LastTransitionTime = metav1.NewTime(time.Now())
+	}
+
+	existingCondition.Reason = newCondition.Reason
+	existingCondition.Message = newCondition.Message
+}
+
+// FindWLPStatusCondition is used to find conditions from config node object status.workerLatencyProfileStatus
+func findWLPStatusCondition(conditions []metav1.Condition, conditionType string) *metav1.Condition {
+	for i := range conditions {
+		if conditions[i].Type == conditionType {
+			return &conditions[i]
+		}
+	}
+
+	return nil
+}
+
+func (c *latencyProfileController) updateConfigNodeStatus(newCondition metav1.Condition) (updated bool, err error) {
+	err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		configNodeObj, err := c.configNodeLister.Get("cluster")
+		if err != nil {
+			updated = false
+			return err
+		}
+		oldStatus := configNodeObj.Status
+
+		newStatus := oldStatus.DeepCopy()
+		setWLPStatusCondition(&newStatus.WorkerLatencyProfileStatus.Conditions, newCondition)
+
+		if equality.Semantic.DeepEqual(oldStatus, newStatus) {
+			updated = false
+			// We return the newStatus which is a deep copy of oldStatus but with set condition applied.
+			return nil
+		}
+
+		configNodeObj.Status = *newStatus
+		_, err = c.configClient.Nodes().UpdateStatus(context.TODO(), configNodeObj, metav1.UpdateOptions{})
```

Comment:
if you want to set the status for a different operator then please move this controller to this operator. This operator should only manage its own status.

Reply:
Actually, we don't have a separate operator for config/v1/node, and if you check the [worker latency profile EP](https://github.com/openshift/enhancements/blob/master/enhancements/worker-latency-profile/worker-latency-profile.md#workerlatencyprofile) the idea is that the respective operand's (operand being Kubelet, Kube Controller Manager, Kube API Server) operator should update the config node status to propagate information about the latency profile rollout.

For now, we have [kubelet's latency profile status being updated by MCO](https://github.com/openshift/machine-config-operator/pull/3015) on the config node object.

Reply:
since the resource isn't managed by an operator what is the point of updating it? The status won't be visible by an end-user.

Reply:
i do agree with @p0lyn0mial that updating node status from KAS operator looks like an anti-pattern to what we do in operators. only kubelet should update the node status, having multiple actors updating this status might only lead to problems and conflicts.

Reply:
nit: please wire the context.TODO() properly :-)

Reply:
Using function signature as `func (c *latencyProfileController) updateConfigNodeStatus(ctx context.Context, newCondition metav1.Condition) (updated bool, err error) {` now to avoid using `context.TODO()`.

Reply:
@p0lyn0mial 
The status is visible to end user via the following command which prints the relevant status from the config node object.
```
$ oc get node.config -o yaml
  status:
    workerLatencyProfileStatus:
      conditions:
      - lastTransitionTime: "2022-04-25T09:28:02Z"
        message: worker latency profile not set on cluster
        reason: ProfileEmpty
        status: "False"
        type: KubeControllerManagerDegraded
      - lastTransitionTime: "2022-04-25T10:03:51Z"
        message: Machine Config Pool Updated
        reason: MachineConfigPoolUpdated
        status: "True"
        type: KubeletComplete
      - lastTransitionTime: "2022-04-25T10:03:50Z"
        message: kube-api-server(s) are updating the latency profile
        reason: ProfileUpdated
        status: "True"
        type: KubeAPIServerProgressing
```

@mfojtik 
For this feature to work we essentially need the args for Kubelet, KubeAPIServer and KubeControllerManager to be updated.
- In case of Kubelet, MCO updates the arg(s)
- In case of APIServer, this operator i.e. KubeAPIServer Operator should update the args(s)
- In case of ControllerManager, KubeControllerManager Operator should update the args(s)

This is documented in detail in the [Worker Latency Profile EP](https://github.com/openshift/enhancements/blob/master/enhancements/worker-latency-profile/worker-latency-profile.md) after discussing with @soltysh, @sttts and @deads2k.

Cc: @harche @rphillips

Reply:
Another example of how multiple operators let user know about the progress of the worker latency profile roll out. 

```
spec:
  workerLatencyProfile: Default
status:
  workerLatencyProfileStatus:
    conditions:
    - lastTransitionTime: "2022-04-26T16:09:39Z"
      message: ""
      reason: ProfileUpdated
      status: "False"
      type: KubeControllerManagerDegraded
    - lastTransitionTime: "2022-04-26T18:30:24Z"
      message: Machine Config Pool Updated
      reason: MachineConfigPoolUpdated
      status: "True"
      type: KubeletComplete
    - lastTransitionTime: "2022-04-26T17:26:38Z"
      message: ""
      reason: ProfileUpdated
      status: "False"
      type: KubeAPIServerProgressing
    - lastTransitionTime: "2022-04-26T16:57:06Z"
      message: all kube-api-server(s) have updated latency profile
      reason: ProfileUpdated
      status: "True"
      type: KubeAPIServerComplete
    - lastTransitionTime: "2022-04-26T18:24:37Z"
      message: ""
      reason: ProfileUpdated
      status: "False"
      type: KubeControllerManagerProgressing
    - lastTransitionTime: "2022-04-26T18:24:37Z"
      message: all kube-controller-manager(s) have updated latency profile
      reason: ProfileUpdated
      status: "True"
      type: KubeControllerManagerComplete
```

Reply:
We're updating status in the operator now 🎉.